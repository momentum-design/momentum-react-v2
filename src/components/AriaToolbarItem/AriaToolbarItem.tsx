import { Props } from './AriaToolbarItem.types';

import React, { forwardRef, useCallback } from 'react';

import { useKeyboard } from '@react-aria/interactions';
import { useAriaToolbarContext } from '../AriaToolbar/AriaToolbar.utils';

const AriaToolbarItem = forwardRef<HTMLButtonElement, Props>((props, providedRef) => {
  const { children, itemIndex } = props;

  const {
    currentFocus,
    setCurrentFocus,
    buttonRefs,
    orientation,
    onTabPress,
    ariaToolbarItemsSize,
  } = useAriaToolbarContext();

  const { keyboardProps } = useKeyboard({
    onKeyDown: (e) => {
      // for the escape key (and other key presses), continue propagation to let Popovers / Modals know that
      // they should close
      e.continuePropagation();

      switch (e.key) {
        case orientation === 'horizontal' ? 'ArrowLeft' : 'ArrowUp':
          e.preventDefault();
          setCurrentFocus((ariaToolbarItemsSize + (currentFocus || 0) - 1) % ariaToolbarItemsSize);
          break;

        case orientation === 'horizontal' ? 'ArrowRight' : 'ArrowDown':
          e.preventDefault();
          setCurrentFocus((ariaToolbarItemsSize + (currentFocus || 0) + 1) % ariaToolbarItemsSize);
          break;

        case 'Tab': {
          if (onTabPress) {
            onTabPress(e);
          }
          break;
        }

        default:
          break;
      }
    },
  });

  const getPropsForChildren = useCallback(
    (child, index) => {
      return {
        tabIndex: index === (currentFocus || 0) ? 0 : -1,
        ref: (e: HTMLButtonElement) => {
          buttonRefs.current[index] = e;
          if (providedRef) {
            if (typeof providedRef === 'function') {
              providedRef(e);
            }
          }
        },
        onFocus: (e) => {
          setCurrentFocus(index);
          if (child.props.onFocus) {
            child.props.onFocus(e);
          }
        },
        onPress: () => {
          setCurrentFocus(index);
          if (child.props.onPress) {
            child.props.onPress();
          }
        },
        useNativeKeyDown: true,
        ...keyboardProps,
      };
    },
    [currentFocus]
  );

  return React.cloneElement(children, getPropsForChildren(children, itemIndex));
});

export default AriaToolbarItem;
