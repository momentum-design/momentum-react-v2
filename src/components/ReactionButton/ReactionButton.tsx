import React, { FC, RefObject, forwardRef } from 'react';
import classnames from 'classnames';
import ButtonCircle, { ButtonCircleSize } from '../ButtonCircle';

import { DEFAULTS, STYLE } from './ReactionButton.constants';
import { Props } from './ReactionButton.types';
import './ReactionButton.style.scss';

/**
 * Button within the ReactionPicker
 */
const ReactionButton = forwardRef((props: Props, ref: RefObject<HTMLButtonElement>) => {
  const { className, children, id, reacted, style, ...otherProps } = props;
  delete otherProps.size;
  return (
    <ButtonCircle
      ref={ref}
      className={classnames(className, STYLE.wrapper)}
      data-reacted={reacted || DEFAULTS.REACTED}
      id={id}
      size={DEFAULTS.SIZE as ButtonCircleSize}
      style={style}
      ghost
      {...otherProps}
    >
      {children}
    </ButtonCircle>
  );
});

ReactionButton.displayName = 'ReactionButton';

export default ReactionButton;
