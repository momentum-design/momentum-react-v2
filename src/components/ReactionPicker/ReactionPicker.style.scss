.md-reaction-picker-wrapper {
  height: 2.25rem;
  background-color: rgba(0, 0, 0, 1);
  background-color: var(--mds-color-theme-groupedbackground-primary-normal);
  border: 0.0625rem solid rgba(255, 255, 255, 0.2);
  border: 0.0625rem solid var(--mds-color-theme-outline-secondary-normal);
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 0, 0.16), 0 0 0.0625rem rgba(0, 0, 0, 0.16);
  overflow: visible;
  display: flex;
  align-items: center;
  gap: 0.5rem;
  justify-content: center;
  width: fit-content;
  margin: 0 0.125rem;
  border-radius: 100vh;
}
