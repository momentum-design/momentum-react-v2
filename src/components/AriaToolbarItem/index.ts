import { default as AriaToolbarItem } from './AriaToolbarItem';
import { Props } from './AriaToolbarItem.types';

export type AriaToolbarItemProps = Props;

export default AriaToolbarItem;
