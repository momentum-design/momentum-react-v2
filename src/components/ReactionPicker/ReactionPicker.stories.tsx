import React from 'react';
import { MultiTemplate, Template } from '../../storybook/helper.stories.templates';
import { DocumentationPage } from '../../storybook/helper.stories.docs';
import StyleDocs from '../../storybook/docs.stories.style.mdx';

import ReactionPicker, { ReactionPickerProps } from './';
import Reaction from '../Reaction';
import ReactionButton from '../ReactionButton';
import argTypes from './ReactionPicker.stories.args';
import Documentation from './ReactionPicker.stories.docs.mdx';

const reactionChildren = [
  <ReactionButton key="1">
    <Reaction name="celebrate" autoPlay />
  </ReactionButton>,
  <ReactionButton key="2">
    <Reaction name="heart" autoPlay />
  </ReactionButton>,
  <ReactionButton key="3">
    <Reaction name="thumb_up_yellow" autoPlay />
  </ReactionButton>,
  <ReactionButton reacted key="4">
    <Reaction name="smile" autoPlay />
  </ReactionButton>,
  <ReactionButton key="5">
    <Reaction name="haha" autoPlay />
  </ReactionButton>,
  <ReactionButton key="6">
    <Reaction name="wow" autoPlay />
  </ReactionButton>,
  <ReactionButton key="7">
    <Reaction name="sad" autoPlay />
  </ReactionButton>,
];

export default {
  title: 'Momentum UI/ReactionPicker',
  component: ReactionPicker,
  parameters: {
    expanded: true,
    docs: {
      page: DocumentationPage(Documentation, StyleDocs),
    },
  },
  subComponents: { ReactionButton },
  args: {
    children: reactionChildren,
  },
};

const Example = Template<ReactionPickerProps>(ReactionPicker).bind({});

Example.argTypes = { ...argTypes };

const Common = MultiTemplate<ReactionPickerProps>(ReactionPicker).bind({});

Common.argTypes = { ...argTypes };
delete Common.argTypes.children;

Common.parameters = {
  variants: [
    { children: reactionChildren },
    { children: reactionChildren },
    { children: reactionChildren },
  ],
};

export { Example, Common };
