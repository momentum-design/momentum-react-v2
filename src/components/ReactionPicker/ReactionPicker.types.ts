import { CSSProperties, ReactElement } from 'react';

import { ReactionButtonProps } from '../ReactionButton';
import { AriaToolbarProps } from '../AriaToolbar';

export type SupportedChildren = ReactionButtonProps;
export type Props = AriaToolbarProps
