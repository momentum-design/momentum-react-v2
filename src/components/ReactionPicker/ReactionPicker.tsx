import React, { FC, useRef } from 'react';
import classnames from 'classnames';

import { STYLE } from './ReactionPicker.constants';
import { Props } from './ReactionPicker.types';
import './ReactionPicker.style.scss';
import AriaToolbar from '../AriaToolbar';

const ReactionPicker: FC<Props> = (props: Props) => {
  const { children, className, id, style, ariaLabel, ...other } = props;

  return (
    <AriaToolbar
      ariaLabel={ariaLabel}
      className={classnames(className, STYLE.wrapper)}
      id={id}
      style={style}
      orientation="horizontal"
      {...other}
    >
      {children}
    </AriaToolbar>
  );
};

ReactionPicker.displayName = 'ReactionPicker';

export default ReactionPicker;
