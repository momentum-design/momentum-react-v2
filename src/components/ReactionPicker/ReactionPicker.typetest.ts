import { Expect, ExpectExtends, ExpectFalse } from '../../utils/typetest.util';
import { Props } from './ReactionPicker.types';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
type cases = [
  Expect<ExpectExtends<Props, { ariaToolbarItemsSize: 1; 'aria-label': 'Some label' }>>,
  Expect<ExpectExtends<Props, { ariaToolbarItemsSize: 1; 'aria-labelledby': 'some-id' }>>,
  Expect<
    ExpectExtends<
      Props,
      { ariaToolbarItemsSize: 1; 'aria-label': 'Some label'; 'aria-labelledby': 'some-id' }
    >
  >,

  ExpectFalse<ExpectExtends<Props, { 'aria-label': 'Some label' }>>,
  ExpectFalse<ExpectExtends<Props, { 'aria-labelledby': 'some-id' }>>,
  ExpectFalse<ExpectExtends<Props, { 'aria-label': 'Some label'; 'aria-labelledby': 'some-id' }>>,

  ExpectFalse<ExpectExtends<Props, { ariaToolbarItemsSize: 1 }>>,

  ExpectFalse<ExpectExtends<Props, Record<string, never>>>
];
