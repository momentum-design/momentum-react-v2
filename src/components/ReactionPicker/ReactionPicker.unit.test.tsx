import React from 'react';
import { mount } from 'enzyme';

import ReactionPicker, { REACTION_PICKER_CONSTANTS as CONSTANTS } from './';
import ReactionButton from '../ReactionButton';
import AriaToolbarItem from '../AriaToolbarItem';

describe('<ReactionPicker />', () => {
  describe('snapshot', () => {
    it('should match snapshot', () => {
      expect.assertions(1);

      const container = mount(
        <ReactionPicker ariaToolbarItemsSize={0} ariaLabel="test-aria-label" />
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with className', () => {
      expect.assertions(1);

      const className = 'example-class';

      const container = mount(
        <ReactionPicker
          className={className}
          ariaToolbarItemsSize={0}
          ariaLabel="test-aria-label"
        />
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with id', () => {
      expect.assertions(1);

      const id = 'example-id';

      const container = mount(
        <ReactionPicker id={id} ariaToolbarItemsSize={0} ariaLabel="test-aria-label" />
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with style', () => {
      expect.assertions(1);

      const style = { color: 'pink' };

      const container = mount(
        <ReactionPicker style={style} ariaToolbarItemsSize={0} ariaLabel="test-aria-label" />
      );

      expect(container).toMatchSnapshot();
    });

    it('should match snapshot with children', () => {
      expect.assertions(1);

      const children = (
        <AriaToolbarItem itemIndex={0}>
          <ReactionButton />
        </AriaToolbarItem>
      );

      const container = mount(
        <ReactionPicker ariaToolbarItemsSize={1} ariaLabel="test-aria-label">
          {children}
        </ReactionPicker>
      );

      expect(container).toMatchSnapshot();
    });
  });

  describe('attributes', () => {
    it('should have its wrapper class', () => {
      expect.assertions(1);

      const element = mount(<ReactionPicker ariaToolbarItemsSize={0} ariaLabel="test-aria-label" />)
        .find(ReactionPicker)
        .getDOMNode();

      expect(element.classList.contains(CONSTANTS.STYLE.wrapper)).toBe(true);
    });

    it('should have provided class when className is provided', () => {
      expect.assertions(1);

      const className = 'example-class';

      const element = mount(
        <ReactionPicker
          className={className}
          ariaToolbarItemsSize={0}
          ariaLabel="test-aria-label"
        />
      )
        .find(ReactionPicker)
        .getDOMNode();

      expect(element.classList.contains(className)).toBe(true);
    });

    it('should have provided id when id is provided', () => {
      expect.assertions(1);

      const id = 'example-id';

      const element = mount(
        <ReactionPicker id={id} ariaToolbarItemsSize={0} ariaLabel="test-aria-label" />
      )
        .find(ReactionPicker)
        .getDOMNode();

      expect(element.id).toBe(id);
    });

    it('should have provided style when style is provided', () => {
      expect.assertions(1);

      const style = { color: 'pink' };
      const styleString = 'color: pink;';

      const element = mount(
        <ReactionPicker style={style} ariaToolbarItemsSize={0} ariaLabel="test-aria-label" />
      )
        .find(ReactionPicker)
        .getDOMNode();

      expect(element.getAttribute('style')).toBe(styleString);
    });
  });
});
