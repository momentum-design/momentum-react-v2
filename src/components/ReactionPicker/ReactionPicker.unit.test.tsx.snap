// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ReactionPicker /> snapshot should match snapshot 1`] = `
<ReactionPicker
  ariaLabel="test-aria-label"
  ariaToolbarItemsSize={0}
>
  <AriaToolbar
    ariaLabel="test-aria-label"
    ariaToolbarItemsSize={0}
    className="md-reaction-picker-wrapper"
    orientation="horizontal"
  >
    <div
      aria-label="test-aria-label"
      aria-orientation="horizontal"
      className="md-reaction-picker-wrapper md-aria-toolbar-wrapper"
      role="toolbar"
    />
  </AriaToolbar>
</ReactionPicker>
`;

exports[`<ReactionPicker /> snapshot should match snapshot with children 1`] = `
<ReactionPicker
  ariaLabel="test-aria-label"
  ariaToolbarItemsSize={1}
>
  <AriaToolbar
    ariaLabel="test-aria-label"
    ariaToolbarItemsSize={1}
    className="md-reaction-picker-wrapper"
    orientation="horizontal"
  >
    <div
      aria-label="test-aria-label"
      aria-orientation="horizontal"
      className="md-reaction-picker-wrapper md-aria-toolbar-wrapper"
      role="toolbar"
    >
      <ForwardRef
        itemIndex={0}
      >
        <ReactionButton
          onFocus={[Function]}
          onKeyDown={[Function]}
          onPress={[Function]}
          tabIndex={0}
          useNativeKeyDown={true}
        >
          <ButtonCircle
            className="md-reaction-button-wrapper"
            data-reacted={false}
            ghost={true}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onPress={[Function]}
            size={32}
            tabIndex={0}
            useNativeKeyDown={true}
          >
            <ButtonSimple
              aria-disabled={false}
              className="md-button-circle-wrapper md-reaction-button-wrapper"
              data-color="primary"
              data-disabled={false}
              data-ghost={true}
              data-inverted={false}
              data-multiple-children={false}
              data-outline={false}
              data-reacted={false}
              data-shallow-disabled={false}
              data-size={32}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onPress={[Function]}
              tabIndex={0}
              useNativeKeyDown={true}
            >
              <FocusRing>
                <FocusRing
                  focusClass="md-focus-ring-wrapper children"
                  focusRingClass="md-focus-ring-wrapper children"
                >
                  <button
                    className="md-button-circle-wrapper md-reaction-button-wrapper md-button-simple-wrapper"
                    data-color="primary"
                    data-disabled={false}
                    data-ghost={true}
                    data-inverted={false}
                    data-multiple-children={false}
                    data-outline={false}
                    data-reacted={false}
                    data-shallow-disabled={false}
                    data-size={32}
                    onBlur={[Function]}
                    onClick={[Function]}
                    onDragStart={[Function]}
                    onFocus={[Function]}
                    onKeyDown={[Function]}
                    onKeyUp={[Function]}
                    onMouseDown={[Function]}
                    onMouseEnter={[Function]}
                    onMouseLeave={[Function]}
                    onMouseUp={[Function]}
                    onTouchCancel={[Function]}
                    onTouchEnd={[Function]}
                    onTouchMove={[Function]}
                    onTouchStart={[Function]}
                    type="button"
                  />
                </FocusRing>
              </FocusRing>
            </ButtonSimple>
          </ButtonCircle>
        </ReactionButton>
      </ForwardRef>
    </div>
  </AriaToolbar>
</ReactionPicker>
`;

exports[`<ReactionPicker /> snapshot should match snapshot with className 1`] = `
<ReactionPicker
  ariaLabel="test-aria-label"
  ariaToolbarItemsSize={0}
  className="example-class"
>
  <AriaToolbar
    ariaLabel="test-aria-label"
    ariaToolbarItemsSize={0}
    className="example-class md-reaction-picker-wrapper"
    orientation="horizontal"
  >
    <div
      aria-label="test-aria-label"
      aria-orientation="horizontal"
      className="example-class md-reaction-picker-wrapper md-aria-toolbar-wrapper"
      role="toolbar"
    />
  </AriaToolbar>
</ReactionPicker>
`;

exports[`<ReactionPicker /> snapshot should match snapshot with id 1`] = `
<ReactionPicker
  ariaLabel="test-aria-label"
  ariaToolbarItemsSize={0}
  id="example-id"
>
  <AriaToolbar
    ariaLabel="test-aria-label"
    ariaToolbarItemsSize={0}
    className="md-reaction-picker-wrapper"
    id="example-id"
    orientation="horizontal"
  >
    <div
      aria-label="test-aria-label"
      aria-orientation="horizontal"
      className="md-reaction-picker-wrapper md-aria-toolbar-wrapper"
      id="example-id"
      role="toolbar"
    />
  </AriaToolbar>
</ReactionPicker>
`;

exports[`<ReactionPicker /> snapshot should match snapshot with style 1`] = `
<ReactionPicker
  ariaLabel="test-aria-label"
  ariaToolbarItemsSize={0}
  style={
    {
      "color": "pink",
    }
  }
>
  <AriaToolbar
    ariaLabel="test-aria-label"
    ariaToolbarItemsSize={0}
    className="md-reaction-picker-wrapper"
    orientation="horizontal"
    style={
      {
        "color": "pink",
      }
    }
  >
    <div
      aria-label="test-aria-label"
      aria-orientation="horizontal"
      className="md-reaction-picker-wrapper md-aria-toolbar-wrapper"
      role="toolbar"
      style={
        {
          "color": "pink",
        }
      }
    />
  </AriaToolbar>
</ReactionPicker>
`;
